doc/protocol.md

```diff
@@ -829,6 +829,119 @@ or pipe is done.
 Additionally, a "options" control message may be sent in this channel
 to change the "max-size" options.
 
+Payload: fsinfo
+----------------
+
+This channel is used to report information about a given path, and the file or
+a directory which may reside there, including the possibility of updates.  In
+the case of directories, information can also be reported about the (directly)
+contained entries.
+
+Additional "open" command options should be specified with a channel of
+this payload type:
+
+ * `path` (mandatory): a string
+ * `attrs` (mandatory): a list of strings, the list of attributes to be
+   reported for the requested file.  See below for the possibilities.
+ * `fnmatch`: (optional, default ''): if non-empty, this is a [`fnmatch`-style
```

Comment:
Just to avoid doubt:
 - that default means that if `path` is a directory, this won't report any information about the contents, just the dir itself.
 - if `path` is not a directory, this option gets ignored?

This could be clarified.

Reply:
very precisely: the requested information about the file (whatever it is) is always reported.  Additionally, if `fnmatch` is non-empty, and `path` happens to point to a directory at any given time, and we can open that directory for reading, then there will be an `entries` attribute added: a dictionary mapping filenames to objects describing those files.

If `path` is not a directory, or if `path` can't be opened for reading, there will be no `entries`.  If the directory can be opened for reading but is empty (excluding `.` and `..`) then `entries` will be present with a value of `{}`.

This is very similar to requesting the "content" and "target" attributes.  They are only reported if the file type is correct, otherwise they are missing.

Originally you opted into having "entries" reported by having "entries" in the attr array, which was a more logically consistent API.  I found this to be a bit awkwards in use and implementation though, and it led to a lot of complex argument checking.  The new thought it: if we have to have `fnmatch` anyway, this is a far better way to get the same behaviour.